dotnet/src/Agents/OpenAI/OpenAIAssistantChannel.cs

```diff
@@ -0,0 +1,379 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Runtime.CompilerServices;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure;
+using Azure.AI.OpenAI.Assistants;
+using Microsoft.SemanticKernel.Agents.OpenAI.Extensions;
+using Microsoft.SemanticKernel.ChatCompletion;
+
+namespace Microsoft.SemanticKernel.Agents.OpenAI;
+
+/// <summary>
+/// A <see cref="AgentChannel"/> specialization for use with <see cref="OpenAIAssistantAgent"/>.
+/// </summary>
+internal sealed class OpenAIAssistantChannel(AssistantsClient client, string threadId) : AgentChannel<OpenAIAssistantAgent>
+{
+    private const char FunctionDelimiter = '-';
+
+    private static readonly TimeSpan s_pollingInterval = TimeSpan.FromMilliseconds(500);
+    private static readonly TimeSpan s_pollingBackoff = TimeSpan.FromSeconds(1);
+    private static readonly TimeSpan s_messageSynchronizationDelay = TimeSpan.FromMilliseconds(500);
+
+    private static readonly HashSet<RunStatus> s_pollingStatuses =
+        [
+            RunStatus.Queued,
+            RunStatus.InProgress,
+            RunStatus.Cancelling,
+        ];
+
+    private static readonly HashSet<RunStatus> s_terminalStatuses =
+        [
+            RunStatus.Expired,
+            RunStatus.Failed,
+            RunStatus.Cancelled,
+        ];
+
+    private readonly AssistantsClient _client = client;
+    private readonly string _threadId = threadId;
+    private readonly Dictionary<string, ToolDefinition[]> _agentTools = [];
+    private readonly Dictionary<string, string> _agentNames = []; // Cache agent names by their identifier for GetHistoryAsync()
+
+    /// <inheritdoc/>
+    protected override async Task ReceiveAsync(IReadOnlyList<ChatMessageContent> history, CancellationToken cancellationToken)
+    {
+        foreach (var message in history)
+        {
+            if (string.IsNullOrWhiteSpace(message.Content))
+            {
+                continue;
+            }
+
+            // History is only be user or assistant (never system)
+            MessageRole role = message.Role == AuthorRole.User ? MessageRole.User : MessageRole.Assistant;
```

Comment:
If the message has a system role (due to some error) when it will be silently converted to an assistant role. Is this desired?

What about an extension method so we can do `message.Role.ToMessageRole()`

Reply:
For the system role case, `AgentChat` explicitly disallows this at the framework level...for the current code-state, this could be commented for clarity.

In considering the case were it to be encountered, I'm not sure converting to assistant message or throwing would be preferable.  Out of those options, I suppose converting to assistant message is the least disruptive.

For Assistant API, messages with system role never exist on the thread.

I'll ponder this more as I refactor the extension method.

Reply:
Updated.  Added extension method and clarified logic.